service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt

```diff
@@ -351,6 +353,245 @@ class AttendanceReservationController(
+        return db.connect { dbc ->
+                dbc.read { tx ->
+                    ac.requirePermissionFor(
+                        tx,
+                        user,
+                        clock,
+                        Action.Unit.READ_CHILD_RESERVATIONS,
+                        unitId
+                    )
+
+                    val rowsByDate =
+                        tx.getChildReservationsOfUnitForDay(unitId = unitId, day = examinationDate)
+                    val childMap = mutableMapOf<ChildId, ReservationChildInfo>()
+
+                    val childReservationInfos =
+                        rowsByDate
+                            .groupBy { it.childId }
+                            .map { row ->
+
+                                // every row duplicates full basic info for child
+                                val basicInfoItem = row.value[0]
+                                childMap.putIfAbsent(
+                                    row.key,
+                                    ReservationChildInfo(
+                                        id = basicInfoItem.childId,
+                                        firstName = basicInfoItem.firstName,
+                                        lastName = basicInfoItem.lastName,
+                                        preferredName = basicInfoItem.preferredName,
+                                        dateOfBirth = basicInfoItem.dateOfBirth
+                                    )
+                                )
+
+                                // repeating child rows are for separate reservation times
+                                val res =
+                                    row.value
+                                        .sortedBy { it.reservationStartTime }
+                                        .map {
+                                            if (
+                                                it.reservationStartTime != null &&
+                                                    it.reservationEndTime != null
+                                            )
+                                                Reservation.Times(
+                                                    startTime = it.reservationStartTime,
+                                                    endTime = it.reservationEndTime
+                                                )
+                                            else Reservation.NoTimes
+                                        }
+
+                                ChildReservationInfo(
+                                    reservations = res,
+                                    absent = basicInfoItem.absenceType != null,
+                                    groupId = basicInfoItem.groupId,
+                                    childId = basicInfoItem.childId,
+                                    outOnBackupPlacement = basicInfoItem.backupUnitId != null,
+                                    dailyServiceTimes =
```

Comment:
You should at least use a separate `tx.getDailyServiceTimesForChildren()` function call to get daily service times and not have the 50 lines of logic duplicated here.

Reply:
Daily service time logic separated from the main query as proposed.